File: src/IceBuffers.cs
using System;

// IceBuffers.cs -- the streaming engine's ring / plane / staging buffers, owned by openICE so the front and
// back half run entirely against managed memory.
//
// Each ring is a single block sliced into `count` sub-buffers of `width * elemFloats` floats
// (ecc = width + 8 header slop; W = image width). The dimensions per ring are listed inline below.
//
// Ring model: each ring is float[slots][len]. Per row, IceRow.SlotAdvance picks which slot each of the ~12
// rows in the working window maps to (mod the slot count); the builders read/write those slots.

public sealed class IceBuffers
{
    public readonly int W, ecc, ec8;

    // object rings (float[slot][col]), shown as (count, width, elemFloats):
    public readonly float[][] inR, inG, inB;   // (12, ecc, 1)  density-converted input planes
    public readonly float[][] gateHist;        // (11, ecc, 1)  edge-padded gate history
    public readonly float[][] weight;          // (11, ecc, 1)  clamped-linear weight ring
    public readonly float[][] gateCur;         // ( 3, W,   1)  current IR gate from ingest
    public readonly float[][] mask;            // ( 2, W,   4)  confidence mask, 4 floats/col
    public readonly float[][] pyr;             // ( 3, W+2, 4)  gate pyramid, 4 floats/col
    public readonly float[][] outR, outG, outB;// ( 3, ecc, 1)  reconstructed output rows
    public readonly float[][] L0, L1, L2, L3;  // ( 1, W, 3)  reconstruction L-planes

    // static staging (separate from the object rings): pyramid staging 12 slots, product staging 11 slots,
    // plus the two 5-tall horizontal-sum ladders (weight partials and product partials).
    public readonly float[][] pyrStage;        // 12 x 36000 floats (3/col)
    public readonly float[][] prodStage;       // 11 x 12000 floats
    public readonly float[][] weightLadder, productLadder;  // the two 5-tall horizontal-sum ladders (5 x ecc each)

    // one ring: `count` slots, each `width * elemFloats` floats.
    static float[][] Ring(int count, int width, int elemFloats)
    {
        var a = new float[count][];
        for (int i = 0; i < count; i++) a[i] = new float[width * elemFloats];
        return a;
    }

    public IceBuffers(int w)
    {
        W = w; ecc = w + 8; ec8 = 4;
        // --- the 15 object rings (count, width, elemFloats) ---
        inR = Ring(12, ecc, 1); inG = Ring(12, ecc, 1); inB = Ring(12, ecc, 1);
        gateHist = Ring(11, ecc, 1);
        weight   = Ring(11, ecc, 1);
        gateCur  = Ring(3, W, 1);
        mask     = Ring(2, W, 4);
        pyr      = Ring(3, W + 2, 4);
        outR = Ring(3, ecc, 1); outG = Ring(3, ecc, 1); outB = Ring(3, ecc, 1);
        L0 = Ring(1, W, 3); L1 = Ring(1, W, 3); L2 = Ring(1, W, 3); L3 = Ring(1, W, 3);
        // --- static staging (separate alloc site) ---
        pyrStage = Ring(12, 36000, 1);
        prodStage = Ring(11, 12000, 1);
        weightLadder = Ring(5, ecc, 1); productLadder = Ring(5, ecc, 1);
    }

    public long TotalBytes()
    {
        long t = 0;
        Action<float[][]> add = r => { foreach (var s in r) t += (long)s.Length * 4; };
        add(inR); add(inG); add(inB); add(gateHist); add(weight); add(gateCur); add(mask); add(pyr);
        add(outR); add(outG); add(outB); add(L0); add(L1); add(L2); add(L3);
        add(pyrStage); add(prodStage); add(weightLadder); add(productLadder);
        return t;
    }
}

File: src_C/ice_buffers.c
/* ice_buffers.c -- the streaming engine's ring / plane / staging buffers. Ported from IceBuffers.cs. */
#include "ice.h"
#include <stdlib.h>

static float **ring(int count, int width, int elemFloats) {
    float **a = (float **)malloc(count * sizeof(float *));
    int i; for (i = 0; i < count; i++) a[i] = (float *)calloc((size_t)width * elemFloats, sizeof(float));
    return a;
}
static void ring_free(float **a, int count) { int i; if (!a) return; for (i = 0; i < count; i++) free(a[i]); free(a); }

IceBuffers *ice_buffers_create(int W) {
    IceBuffers *b = (IceBuffers *)calloc(1, sizeof(IceBuffers));
    int ecc = W + 8;
    b->W = W; b->ecc = ecc; b->ec8 = 4;
    b->inR = ring(12, ecc, 1); b->inG = ring(12, ecc, 1); b->inB = ring(12, ecc, 1);
    b->gateHist = ring(11, ecc, 1);
    b->weight   = ring(11, ecc, 1);
    b->gateCur  = ring(3, W, 1);
    b->mask     = ring(2, W, 4);
    b->pyr      = ring(3, W + 2, 4);
    b->outR = ring(3, ecc, 1); b->outG = ring(3, ecc, 1); b->outB = ring(3, ecc, 1);
    b->L0 = ring(1, W, 3); b->L1 = ring(1, W, 3); b->L2 = ring(1, W, 3); b->L3 = ring(1, W, 3);
    b->pyrStage = ring(12, 36000, 1);
    b->prodStage = ring(11, 12000, 1);
    b->weightLadder = ring(5, ecc, 1); b->productLadder = ring(5, ecc, 1);
    return b;
}

void ice_buffers_free(IceBuffers *b) {
    if (!b) return;
    ring_free(b->inR, 12); ring_free(b->inG, 12); ring_free(b->inB, 12);
    ring_free(b->gateHist, 11); ring_free(b->weight, 11);
    ring_free(b->gateCur, 3); ring_free(b->mask, 2); ring_free(b->pyr, 3);
    ring_free(b->outR, 3); ring_free(b->outG, 3); ring_free(b->outB, 3);
    ring_free(b->L0, 1); ring_free(b->L1, 1); ring_free(b->L2, 1); ring_free(b->L3, 1);
    ring_free(b->pyrStage, 12); ring_free(b->prodStage, 11);
    ring_free(b->weightLadder, 5); ring_free(b->productLadder, 5);
    free(b);
}
